### ISSUE TITLE
Log `Record`s passed to log processor `OnEmit` is non-mutable

### ISSUE THREAD

**Issue by MrAlias** [open-telemetry/opentelemetry-go#5219]
Blocked by:

- https://github.com/open-telemetry/opentelemetry-specification/issues/4065

### Description

[The specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/sdk.md#onemit) requires that log processors be provided a `LogRecordProcessor`:

> #### OnEmit
> 
> [...]
> 
> **Parameters:**
> 
> * `logRecord` - a [ReadWriteLogRecord](#readwritelogrecord) for the emitted `LogRecord`.

[Our current implementation](https://github.com/open-telemetry/opentelemetry-go/blob/v1.25.0/sdk/log/processor.go) Defines the `Processor` interface as such:

```go
type Processor interface {
	// OnEmit is called when a Record is emitted.
	//
	// OnEmit will be called independent of Enabled. Implementations need to
	// validate the arguments themselves before processing.
	//
	// Implementation should not interrupt the record processing
	// if the context is canceled.
	//
	// All retry logic must be contained in this function. The SDK does not
	// implement any retry logic. All errors returned by this function are
	// considered unrecoverable and will be reported to a configured error
	// Handler.
	//
	// Before modifying a Record, the implementation must use Record.Clone
	// to create a copy that shares no state with the original.
	OnEmit(ctx context.Context, record Record) error
/* ... */
}
```

Any modification made to the `record` passed to a processor's `OnEmit` method will not be applied to the subsequent processor. The record is not mutable, it does not comply with the specification requirement.

**Comment by MrAlias** [comment]
One solution is to have `OnEmit` accept a reference:

```go
OnEmit(ctx context.Context, record *Record) error
```

This will mean that every record will need to be allocated to the heap.

**Comment by MrAlias** [comment]
Another solution is to have `OnEmit` return a `Record`

```go
OnEmit(ctx context.Context, record Record) (Record, error)
```

The returned `Record` can be the same record passed, or a mutated clone of the `Record`. That returned value will be passed to the next processor in the SDK's processing chain.

**Comment by pellared** [comment]
I could work on this but it would be nice to have https://github.com/open-telemetry/opentelemetry-go/pull/5093 merged first to reduce the amount of conflicts.

**Comment by MrAlias** [comment]
> I could work on this but it would be nice to have #5093 merged first to reduce the amount of conflicts.

Merged

**Comment by pellared** [comment]
I remember what was my idea. It was to use Processor as decorators in order to make modifications.

So e.g this would be a processor that "modifies" the record:

```go
type mutatingProcessor struct {
  wrapped log.Processor
}

func (mutatingProcessor p) OnEmit(ctx context.Context, record Record) error {
  r.SetBody(log.StringValue("REDACTED"))
  return p.wrapped.OnEmit(ctx, r.Clone())
}
```

Thanks to it you can have even two "processor" pipelines which do not interfere with each other.

So you can have one processor pipeline which e.g. has a processor redacts secrets and wraps a simple processor with stdout exporter and then a second pipeline which uses a batch processor (no redacting) with OTLP exporter:

```go

log.NewLoggerProvider(
  log.WithProcessor(mutatingProcessor{ log.NewSimpleProcessor(stdoutExp) }),
  log.WithProcessor(log.NewBatchProcessor(otlpExp)),
)

**Comment by MrAlias** [comment]
No, I don't think we can close the issue. We are not compatible with the specification. The specification states the log records are mutable and that processors are executed in order. The processors after a mutation should receive the mutated record.

**Comment by pellared** [comment]
> The processors after a mutation should receive the mutated record.

Where is it defined?

What do you think about updating the specification?

**Comment by MrAlias** [comment]
> Where is it defined?

In the definition of a read/write record.

**Comment by pellared** [comment]
I think it is "in spirit" of the specification. I would rather ask the Specification SIG if it is acceptable design and (optionaly) clarify the specification.

**Comment by pellared** [comment]
I created https://github.com/open-telemetry/opentelemetry-specification/issues/4010

**Comment by pellared** [comment]
It looks that the TC wants to have a possibility so that the mutated record will be passed to next registered processor without forcing the user to make a decorator.

See: https://github.com/open-telemetry/opentelemetry-specification/pull/4062/files#r1619504122

PS. There is a chance that making `OnEmit(ctx context.Context, record *Record) error` would not cause any performance overhead in "regular" use case.

**Comment by pellared** [comment]
I created https://github.com/open-telemetry/opentelemetry-specification/issues/4065
